### PR TITLE
Add ConstantTrhoDomain for fragment-based film growth simulation

### PR DESCRIPTION
**Pull request by hwpang**

### CODE REVIEW

File: src/Reactor.jl [hunk 1]
```diff
@@ -471,6 +472,53 @@ export getrates
     end
 end
 
+function addreactionratecontributions!(dydt::Q,fragmentbasedrxnarray::Array{W2,2},rxnarray::Array{Int64,2},cs::W,kfs::Z,krevs::Y,massindex::Int64,Mws::Array{Float64,1},fragmentindexes::Array{Int64,1}) where {Q,Z,Y,T,W,W2}
+    (numspcs,numrxns) = size(fragmentbasedrxnarray)
+    half = Int(numspcs/2)
```
Comment (mjohnson541) [inline src/Reactor.jl:477]: Can you improve the variable name for `numspcs`?

File: src/Domain.jl [hunk 2]
```diff
@@ -703,6 +703,121 @@ function ConstantTAPhiDomain(;phase::E2,initialconds::Dict{X,X2},constantspecies
 end
 export ConstantTAPhiDomain
 
+mutable struct ConstantTrhoDomain{N<:AbstractPhase,S<:Integer,W<:Real, W2<:Real, I<:Integer, Q<:AbstractArray} <: AbstractConstantKDomain
```
Comment (mjohnson541) [inline src/Domain.jl:706]: Can we make it obvious in the name that it is fragment specific?

File: src/Reactor.jl [hunk 3]
```diff
@@ -61,7 +61,8 @@ function Reactor(domain::T,y0::Array{T1,1},tspan::Tuple,interfaces::Z=[];p::X=Sc
         recsolver = Sundials.CVODE_BDF(linear_solver=:GMRES)
     else
         ode = ODEProblem(odefcn,y0,tspan,p)
-        if sparsity > 0.8 #empirical threshold to use preconditioner
+        #empirical threshold to use preconditioner
+        if sparsity > 0.8  && domain isa Union{ConstantTPDomain,ConstantVDomain,ConstantPDomain,ParametrizedTPDomain,ParametrizedVDomain,ParametrizedPDomain,ConstantTVDomain,ParametrizedTConstantVDomain,ConstantTAPhiDomain}
```
Comment (mjohnson541) [inline src/Reactor.jl:65]: Maybe don't need now?

File: src/Simulation.jl [hunk 4]
```diff
@@ -230,17 +232,43 @@ this outputs a sparse matrix of  num reactions xnum species containing the produ
 rate of that species associated with that reaction
 """
 function rops(bsol::Q,t::X) where {Q<:Simulation,X<:Real}
-    ropmat = spzeros(length(bsol.domain.phase.reactions),length(bsol.domain.phase.species))
```
Comment (mjohnson541) [inline src/Simulation.jl:233]: Check that it's not necessary to update other simulation post-processing.

File: src/Domain.jl [hunk 5]
```diff
@@ -703,6 +703,121 @@ function ConstantTAPhiDomain(;phase::E2,initialconds::Dict{X,X2},constantspecies
...
+    thermovariabledict::Dict{String,Int64}
+    Mws::Array{Float64,1}
+    fragmentindexes::Array{Int64,1}
+    fragmentbasedspcnames::Array{String,1}
+    spcnames::Array{String,1}
+    epsilon::Float64
+    diffusionlength::Float64
+end
+
+function ConstantTrhoDomain(;phase::Z,initialconds::Dict{X,E},fragmentnames::Array{X3,1},fragment_based_reaction_mapping::Dict{X1,E1},epsilon::Float64=1.0,diffusionlength::Float64=Inf,constantspecies::Array{X4,1}=Array{String,1}(),
```
Comment (mjohnson541) [inline src/Domain.jl:738]: I think we can stick this information on the Species and Reaction objects. Can add attributes to species flagging whether a Species is a fragment or fragmentintermediate.

Comment (mjohnson541) [inline src/Domain.jl:738]: epsilon => porosity

Comment (hwpang) [reply src/Domain.jl]: Done

Comment (hwpang) [reply src/Domain.jl]: Done

File: src/Simulation.jl (hunk 4 above)
Comment (hwpang) [reply src/Simulation.jl:233]: Done

File: src/Reactor.jl (hunk 3 above)
Comment (hwpang) [reply src/Reactor.jl]: Removed

File: src/Domain.jl (hunk 2 above)
Comment (hwpang) [reply src/Domain.jl]: Done

File: src/Reactor.jl (hunk 1 above)
Comment (hwpang) [reply src/Reactor.jl]: Done

**Review by mjohnson541 — APPROVED** [review]

**Pull request merged**